rust/kernel/init.rs

```diff
@@ -0,0 +1,827 @@
+/// ```rust
+/// # #![allow(clippy::blacklisted_name, clippy::new_ret_no_self)]
+/// # use kernel::{init, pin_init, pin_data, init::*};
+/// # use core::pin::Pin;
+/// # pin_data! { struct Foo {
+/// #     a: usize,
+/// #     b: Bar,
+/// # }}
+/// # pin_data! { struct Bar {
+/// #     x: u32,
+/// # }}
+/// # impl Foo {
+/// #     fn new() -> impl PinInit<Self> {
+/// #         pin_init!(Self {
+/// #             a: 42,
+/// #             b: Bar {
+/// #                 x: 64,
+/// #             },
+/// #         })
+/// #     }
+/// # }
+/// pin_data! {
+///     struct FooContainer {
+///         #[pin]
+///         foo1: Foo,
+///         #[pin]
+///         foo2: Foo,
+///         other: u32,
+///     }
+/// }
+///
+/// impl FooContainer {
+///     fn new(other: u32) -> impl PinInit<Self> {
+///         pin_init!(Self {
+///             foo1: Foo::new(),
+///             foo2: Foo::new(),
+///             other,
+///         })
+///     }
+/// }
+/// ```
+#[macro_export]
+macro_rules! pin_init {
+    ($(&$this:ident in)? $t:ident $(<$($generics:ty),* $(,)?>)? {
+        $($field:ident $(: $val:expr)?),*
+        $(,)?
+    }) => {
+        $crate::pin_init!(@this($($this)?), @type_name($t $($($generics),*)?), @typ($t $($($generics),*)?), @fields($($field $(: $val)?),*))
+    };
+    (@this($($this:ident)?), @type_name($t:ident $(<$($generics:ty),*)?), @typ($ty:ty), @fields($($field:ident $(: $val:expr)?),*)) => {{
+        // we do not want to allow arbitrary returns
+        struct __InitOk;
+        let init = move |slot: *mut $ty| -> ::core::result::Result<__InitOk, _> {
+            {
+                // shadow the structure so it cannot be used to return early
+                struct __InitOk;
+                $(let $this = unsafe { ::core::ptr::NonNull::new_unchecked(slot) };)?
+                $(
+                    $(let $field = $val;)?
+                    // call the initializer
+                    // SAFETY: slot is valid, because we are inside of an initializer closure, we return
+                    //         when an error/panic occurs.
+                    unsafe {
+                        <$ty as $crate::init::__private::__PinData>::__PinData::$field(
+                            ::core::ptr::addr_of_mut!((*slot).$field),
+                            $field,
+                        )?;
+                    }
+                    // create the drop guard
+                    // SAFETY: we forget the guard later when initialization has succeeded.
+                    let $field = unsafe { $crate::init::__private::DropGuard::new(::core::ptr::addr_of_mut!((*slot).$field)) };
+                )*
+                #[allow(unreachable_code, clippy::diverging_sub_expression)]
+                if false {
+                    let _: $t $(<$($generics),*>)? = $t {
+                        $($field: ::core::todo!()),*
+                    };
+                }
+                $(
+                    ::core::mem::forget($field);
```

Comment:
Maybe add a comment that this is the drop guard. At first I got confused and thought it was the actual initialization value.